docs/develop/hw-debugging.md

```diff
@@ -0,0 +1,116 @@
+
+**NOTE:** Currently these docs are written assuming you're running MacOS. They should work with minor modifications on Linux or WSL.
+
+## Prerequisites
+0. Ensure you've reviewed the [pxt README](https://github.com/microsoft/pxt/blob/master/README.md) and [pxt-adafruit README](https://github.com/microsoft/pxt-adafruit/blob/master/README.md) and that you can run pxt-adafruit on localhost with a local pxt-core and pxt-common-packages.
+1. Ensure you have pxt, pxt-common-packages, and pxt-adafruit cloned and sym-linked together (e.g. adafruit points to local pxt/ and common-packages)
+2. Install these things (or equivalent on non-mac):
+    ```
+    brew install openocd
+    brew tap ArmMbed/homebrew-formulae
+    brew install arm-none-eabi-gcc
+    brew install cmake
+    ```
+3. Ensure "openocd" is found at /usr/local/bin/ or /usr/bin/
+4. Run `openocd --version` and ensure you have at least `Open On-Chip Debugger 0.10.0`
+
+## Building a project with local pxt and codal .cpp
+### First time debugging project setup
+1. Build and run pxt-adafruit on localhost and save a project in the web editor (e.g. name it "CPXDebugging")
+2. In a bash shell, navigate to `.../pxt-adafruit/projects/CPXDebugging/`
+3. Set these environment variables (or put them in your .bashrc):
+    ```
+    export PXT_FORCE_LOCAL=1
+    export PXT_NODOCKER=1 
+    export PXT_RUNTIME_DEV=1
+    export PXT_ASMDEBUG=1
+    ```
+    Run `pxt help` for explanations.
+4. Open the "main.ts" file in your project and add this to the top:
+    ```
+    namespace userconfig {
+        export const SWD_ENABLED = 1;
+    }
+    ```
+    This is only needed for CPX and tells it to stop using one of the debugging pins for the amplifier.
+5. In pxt.json, change the dependency to:
+    ```
+    "circuit-playground": "file:../../libs/circuit-playground/"
+    ```
+    This tells the project to use the local circuit-playground package.
+6. Delete your hexcache:
+    ```
+    $ cd ~/mc/pxt-adafruit/built
+    $ rm -rf hexcache/
+    ```
+    This ensure we're building fresh bits.
+
+### Building & Deploying
+1. Ensure CPX is connected to computer by regular micro-usb cable (in addition to whatever debugger connection you may have)
+2. Reset the CPX into flashing mode
+3. From within `.../pxt-adafruit/projects/CPXDebugging` run `pxt`.
+    This should build and deploy to the attached CPX. This builds all our .cpp libs and codal locally and creates a checkout of CODAL to `built/codal/`
+
+## Debugging
+### Connect the debugger HW
+1. Connect the Particle Debugger to the SWD pins on the CircuitPlayground. Probably by:
+    - soldering 3 jumpers to the CPX using this diagram:
+        https://learn.adafruit.com/assets/47156
+        Solder to: SWCCLK (45), SWDIO (46), RESET (40)
+    - connecting those to a SWD breakout board like this:
+        https://www.adafruit.com/product/2743
```

Comment:
Alternatively, connect jumper cables directly to particle debugger. Pins are labelled at the bottom.